src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -547,6 +561,12 @@ internal bool TypeIsCached(Type type)
             return _classes.ContainsKey(type);
         }
 
+        internal void ClearClasses()
```

Comment:
Could custom converters for updated types also be stale here? We might need to also clear the `_converters` cache.

At this point, the options instance is immutable because the first (de)serialization has occurred. If we do need to clear the converters cache, then I believe we'll need to reset `_haveTypesBeenCreated` so that the options instance can be user-modified to replace old converter instances. @steveharter, thoughts?

Reply:
> Could custom converters for updated types also be stale here?

You tell me :smile:  I can clear that as well if it's the right thing to do.

Reply:
In normal cases, `_haveTypesBeenCreated` wouldn't need to be cleared since converters don't hold a reference to the metadata on JsonClassInfo. What is important is that JsonClassInfo metadata cached via reflection\codegen, not the converters themselves.

However since nothing stops a custom converter from caching its own metadata for a class, clearing that variable will allow the converter to be re-created with custom metadata. However, that wouldn't work for explicit user-added converters via options.Converters.Add() since there is no way to re-add those.

If we end up needing to clear that variable, then we'd have to add another variable so that the Options instance isn't mutable temporarily when the caches are flushed via the `VerifyMutable()` method.

Reply:
My goal here is to clear implicit caching the dev has no visibly into.  If the dev is caching their own state,  which it sounds like would be the problem with a converter,  they're on the hook for resetting that via the same update notification mechanism if they care to.